zerver/lib/email_mirror_server.py

```diff
@@ -0,0 +1,190 @@
+                recipient_realm = await sync_to_async(
+                    lambda a: decode_stream_email_address(a)[0].realm
+                )(address)
+                rate_limit_mirror_by_realm(recipient_realm)
+        except RateLimitedError:
+            logger.warning(
+                "Rejecting a MAIL FROM: %s to realm: %s - rate limited.",
+                envelope.mail_from,
+                recipient_realm.name,
+            )
+            return "550 4.7.0 Rate-limited due to too many emails on this realm."
+
+        except ZulipEmailForwardError as e:
+            return f"550 5.1.1 Bad destination mailbox address: {e}"
+
+        envelope.rcpt_tos.append(address)
+        return "250 Continue"
+
+    @override
+    def handle_message(self, message: email.message.Message) -> None:
+        msg_base64 = base64.b64encode(bytes(message))
+
+        for address in message["X-RcptTo"].split(", "):
+            if address == "postmaster":
+                send_to_postmaster(message)
+            else:
+                queue_json_publish_rollback_unsafe(
+                    "email_mirror",
+                    {
+                        "rcpt_to": address,
+                        "msg_base64": msg_base64.decode(),
+                    },
+                )
+
+
+class PermissionDroppingUnthreadedController(UnthreadedController):  # nocoverage
+    @override
+    def __init__(
+        self,
+        user: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        if user is not None:
+            self.user: int | None = pwd.getpwnam(user).pw_uid
+        else:
+            self.user = None
+
+    @override
+    def _create_server(self) -> Awaitable[asyncio.AbstractServer]:
+        # Make the listen socket, then drop privileges before starting
+        # the server
+        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        server_socket.bind((self.hostname, self.port))
+        if self.user and os.geteuid() == 0:
+            logger.info("Dropping privileges to uid %d", self.user)
+            os.setuid(self.user)
+
+        server = self.loop.create_server(
+            self._factory_invoker,
+            sock=server_socket,
+            ssl=self.ssl_context,
+        )
+
+        return server
+
+
+def run_smtp_server(
+    user: str | None, host: str, port: int, tls_context: SSLContext | None
+) -> None:  # nocoverage
+    logger.info("Listening on %s:%d", host, port)
+    server = PermissionDroppingUnthreadedController(
+        user=user,
+        hostname=host,
+        port=port,
+        handler=ZulipMessageHandler(),
+        tls_context=tls_context,
+        ident=f"Zulip {ZULIP_VERSION}",
+    )
+
+    server.loop.add_signal_handler(signal.SIGINT, server.loop.stop)
+
+    server.begin()
+    with suppress(KeyboardInterrupt):
+        server.loop.run_forever()
```

Comment:
Why do we suppress `KeyboardInterrupt` here? Maybe the best answer would be a comment.